### ISSUE TITLE
Add functional analysis code to the main paper

### ISSUE THREAD

**Issue by lpantano** [hbc/bcbioRNASeq#80]
For the QC and differential expression analysis, the article outlines the analysis steps in detail. However, the functional analysis is only described through the existence of an Rmarkdown template. It would be nice to have at least part of the functional analysis also explained and written out in the article.

**Comment by marypipes** [comment]
In GSEA template: remove bitr() sequence and use the Entrez IDs from annotable()

To have one ENSEMBL ID mapping to one ENTREZ ID I did the following:

# Extract the Entrez IDs
allEntrez <- allGenes_ids$entrezid
allEntrez <- unlist(allGenes_ids$entrezid)

# Remove all NA values and only keep unique IDs
allEntrez <- as.character(unique(allEntrez[which(!(is.na(allEntrez)))]))
allEntrez_df <- allGenes_ids[allGenes_ids$entrezid %in% allEntrez, ]

# Combine the results for all genes with the Entrez ID
allEntrez_res <- merge(allEntrez_df, data.frame(res), by.x="gene_id", by.y=0)

# Remove duplicate Entrez IDs
allEntrez_res <- allEntrez_res[which(!(duplicated(allEntrez_res$entrezid))), ]

# Named list of log2 foldchanges
allEntrez_fc <- allEntrez_res$log2FoldChange
names(allEntrez_fc) <- allEntrez_res$entrezid

# Sort by fold change values
allEntrez_fc <- sort(allEntrez_fc, decreasing = TRUE)
sigEntrez <- unlist(allGenes_ids[allGenes_ids$gene_id %in% sigGenes, "entrezid"])
sigEntrez <- as.character(unique(sigEntrez[which(!(is.na(sigEntrez)))]))

kegg <- enrichKEGG(
    gene = sigEntrez,
    universe = allEntrez,
    organism = params$species)

Extra text to add to GSEA:
GSEA analysis is performed with the clusterProfiler tool using KEGG gene sets and using the log2 fold changes as input. By using the log2 fold changes as the input, we are identifying pathways with genes that exhibit coordinated fold changes that are larger than might be expected by chance. The significant pathways can be visualized using the log2 fold changes with the Pathview tool.

Gene set enrichment analysis tools use ranked lists of genes (here ranked by log2FC) without using a threshold. This allows the tools to use more information to identify enriched biological processes. The introduction to gene set enrichment analysis goes into more detail about some of the advantages of this approach: [http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1239896/](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1239896/). By using the log2 fold changes as the input, we are identifying pathways with genes that exhibit coordinated fold changes that are larger than might be expected by chance. The significant pathways can be visualized using the log2 fold changes with the Pathview tool.

The significantly dysregulated pathways (q-value (FDR) < 0.05) are displayed below in the pathway images, which show the degree of dysregulation of the genes (the minus direction (green) is down-regulated, while the positive direction (red) is up-regulated). 

When performing GSEA analysis it may be useful to adjust the `minGSSize` and/or `maxGSSize` parameter based on the pathways you would like to search for significance. If you are interested in smaller pathways, such as phototransduction, which has a gene set size of 24 genes, then you would want to adjust the minGSSize to less than 24. If you are only interested in larger pathways, then you would want to adjust the GSSize to a larger number. The fewer pathways tested, the less we need to correct for multiple test correction, so by adjusting the minGSSize and maxGSSize parameters you can test fewer pathways and limit testing to the pathways of interest.

gseaKEGG <- gseKEGG(
    geneList = allEntrez_fc,
    organism = tolower(params$gspecies),
    nPerm = 1000,
    minGSSize = 20, # or make this default and not include
    pvalueCutoff = 0.05,
    verbose = FALSE)

**Comment by marypipes** [comment]
Also, if possible, can we change in the template the gseGO() to use the ranked fold changes instead of the p-values? Not worth holding up publication, but would be nice to do.

**Comment by mjsteinbaugh** [comment]
@marypiper I agree that we should switch to using the ranked fold changes. Let's update the template on a new branch and do a pull request

**Comment by marypipes** [comment]
One last thing, prior to the GSEA in the functional analysis template, can we add the following lines for the cnetplot:

# Limit the fold change values so that larger fold changes aren't washed out by an outlier that increases the range of FC values
allEntrez_fc <- ifelse(allEntrez_fc > 2, 2, allEntrez_fc)
allEntrez_fc <- ifelse(allEntrez_fc < -2, -2, allEntrez_fc)

**Comment by mjsteinbaugh** [comment]
Thanks @marypiper, I worked on adding these updates into the functional analysis R Markdown template on the develop branch (v0.2.0). I made a few tweaks to the Entrez ID matching method.

R Markdown template link:
https://github.com/hbc/bcbioRNASeq/blob/develop/inst/rmarkdown/templates/functional_analysis/skeleton/skeleton.Rmd

I'm finishing re-running the GSE65267 example dataset with Ensembl GRCm38 release 90 annotations (GRCm38_90 in biodata) on O2 and then I'll post a link to the example bcbioRNASeq object that we can use for testing.
